docs/jmespath-grammar.txt

```diff
@@ -0,0 +1,107 @@
+expr      = ws (terminal / non-terminal) ws
+root-expr = non-test / not / comparison
+non-test  = subexpr / index / flatten / wildcard-index / filter /
+            identifier / current-node / literal / root-multi-list /
+            multi-hash / function / group / slice
+group     = open-group expr close-group
+
+; Insignificant whitespace is allowed around expr nodes and around the
+; following structural tokens.
+open-group      = ws "(" ws
+close-group     = ws ")" ws
+open-bracket    = ws "[" ws
+close-bracket   = ws "]" ws
+open-brace      = ws "{" ws
+close-brace     = ws "}" ws
+dot-separator   = ws "." ws
+pipe-separator  = ws "|" ws
+or-separator    = ws "||" ws
+and-separator   = ws "&&" ws
+name-separator  = ws ":" ws
+value-separator = ws "," ws
+lt              = ws "<" ws
+lte             = ws "<=" ws
+gt              = ws ">" ws
+gte             = ws ">=" ws
+eq              = ws "==" ws
+ne              = ws "!=" ws
+expref-token    = "&" ws
+not-token       = "!" ws
+
+; "&&" binds more tightly than "||"
+; "||" binds more tightly than "|".
+terminal     = pipe / or / and
+non-terminal = root-expr / wildcard-values
+pipe         = expr pipe-separator (non-terminal / or / and)
+or           = (non-terminal / and) or-separator (non-terminal / or / and)
+and          = non-terminal and-separator (non-terminal / and)
+
+; subexpr handles most that descend into nodes.
+subexpr            = object-subexpr / array-subexpr
+object-subexpr     = object-subexpr-lhs dot-separator object-subexpr-rhs
+object-subexpr-lhs = subexpr / index / flatten / wildcard-index /
+                     filter / identifier / current-node / literal /
+                     multi-hash / function / group / slice / wildcard-values
+object-subexpr-rhs = identifier / multi-list / multi-hash / function /
+                     wildcard-values
+array-subexpr      = array-subexpr-lhs array-subexpr-rhs
+array-subexpr-lhs  = subexpr / index / flatten / wildcard-index / filter /
+                     identifier / current-node / literal / root-multi-list /
+                     function / group / slice / wildcard-values / literal
+array-subexpr-rhs  = index / slice / wildcard-index / flatten / filter
+
+; Array related rules
+index           = open-bracket number close-bracket
+slice           = open-bracket *number name-separator
+                               *number *(name-separator *number)
+                               close-bracket
+wildcard-index   = open-bracket "*" close-bracket
+flatten          = open-bracket close-bracket
+filter           = open-bracket "?" filter-condition close-bracket
```

Comment:
I was thinking a single token, i.e `ws "[?" ws`.  It will free up the `?` char if we ever want to use it as something else (like C's a ternary operator).